src/algebra/homology/local_cohomology.lean

```diff
@@ -0,0 +1,173 @@
+Ext into `M`.  One can, of course, take the collection to simply be the integral powers of `I`.
+
+## References
+
+* [M. Hochster, *Local cohomology*][hochsterunpublished]
+  <https://dept.math.lsa.umich.edu/~hochster/615W22/lcc.pdf>
+* [R. Hartshorne, *Local cohomology: A seminar given by A. Grothendieck*][hartshorne61]
+* [M. Brodmann and R. Sharp, *Local cohomology: An algebraic introduction with geometric
+  applications*][brodmannsharp13]
+* [S. Iyengar, G. Leuschke, A. Leykin, Anton, C. Miller, E. Miller, A. Singh, U. Walther,
+  *Twenty-four hours of local cohomology*][iyengaretal13]
+
+## Tags
+
+local cohomology, local cohomology modules
+
+## Future work
+
+* Prove that this definition is equivalent to:
+    * the right-derived functor definition
+    * the characterization as the limit of Koszul homology
+    * the characterization as the cohomology of a Cech-like complex
+* Prove that local cohomology depends only on the radical of the ideal
+* Establish long exact sequence(s) in local cohomology
+-/
+
+open opposite
+open category_theory
+open category_theory.limits
+
+noncomputable theory
+
+universes u v
+
+/- We define local cohomology, implemented as a direct limit of Ext(R/J, --). -/
+section local_cohomology
+
+section
+variables {R : Type u} [comm_ring R] {D : Type v} [small_category D]
+
+local attribute [ext] quiver.hom.unop_inj
+
+/--  The directed system of `R`-modules of the form `R/J`, where `J` is an ideal of `R`,
+determined by the functor `I`  -/
+def ring_mod_ideals (I : D ⥤ ideal R) : D ⥤ Module.{u} R :=
+{ obj := λ t, Module.of R $ R ⧸ (I.obj t),
+  map := λ s t w, submodule.mapq _ _ (linear_map.id) (I.map w).down.down }
+
+/-- The diagram we will take the colimit of to define local cohomology, corresponding to the
+directed system determined by the functor `I` -/
+def local_cohomology_diagram (I : D ⥤ ideal R) (i : ℕ) : Dᵒᵖ ⥤ Module.{u} R ⥤ Module.{u} R :=
+(ring_mod_ideals I).op ⋙ Ext R (Module.{u} R) i
+
+end
+
+section
+-- We momentarily need to work with a type inequality, as later we will take colimits
+-- along diagrams either in Type, or in the same universe as the ring, and we need to cover both.
+variables {R : Type max u v} [comm_ring R] {D : Type v} [small_category D]
+
+/-- `local_cohomology I i` is the the functor sending an module `M` over a commutative ring `R`
+to the direct limit of `Ext^i(R/J, M)`, where `J` ranges over a collection of ideals of `R`
+that is represented as a functor `I`.
+
+In this definition we do not assume any special property of the diagram `I`, but the relevant case
+will be where `I` is cofinal with the diagram of powers of a single given ideal.
+
+Below, we give two equivalent (to be shown) definitions of the usual local cohomology with support
+in an ideal `J`, `local_cohomology_powers` and `local_cohomology_univ`.
+
+TODO: Show that any functor cofinal with `I` gives the same result.
+ -/
```

Comment:
This docstring is probably too long. This is what the user sees when they hover over `local_cohomology` in VS Code, so you can't really talk about things like "below" because it won't be be below. I think it's best to have the docstring just being a succinct accurate description of the definition, and then everything else can be in a comment afterwards.

Another issue I have with this definition is that the usual usage of "local cohomology" in the literature is just the one for powers of `I`, so that one should really be the one called `local_cohomology` I think, rather than this one.
```suggestion
/-- `local_cohomology I i` is the the functor sending an module `M` over a commutative ring `R`
to the direct limit of `Ext^i(R/J, M)`, where `J` ranges over a collection of ideals of `R`
that is represented as a functor `I`.
-/
/-
In this definition we do not assume any special property of the diagram `I`, but the relevant case
will be where `I` is cofinal with the diagram of powers of a single given ideal.

Below, we give two equivalent (to be shown) definitions of the usual local cohomology with support
in an ideal `J`, `local_cohomology_powers` and `local_cohomology_univ`.

TODO: Show that any functor cofinal with `I` gives the same result.
 -/
```

Reply:
Sounds good - I split it up as you suggested.